File: analysis/engine/ratings.py
"""Rating layer (spec §12–§14) — turns lineups into logit / yardage modifiers.

modifier(family) = Σ_attr  beta_per_z(attr) · clip(z(attr), -3, 3)

where z(attr) = (unit_mean_rating - ref_mean) / ref_sd over the participating
players who have that attribute. Signs are already folded into `beta_per_z`
(step 2 of 22_rating_calibration.py).

§12 centering: `ref_mean` is the mean over the whole 1,024-player reference
pool, but the engine only ever fields *starters*, who sit above that mean — so
a raw modifier is non-zero even for a league-average matchup and league
scoring drifts down (seen in the first §23 pass before this fix). Each
`*_shift` therefore subtracts that family's league-mean modifier over its
on-field slot (`_offsets()`), so an average real matchup → ≈0 shift and only
matchup *differences* move the outcome. With this in place the paired ON/OFF
§23 check holds league points flat and widens score-margin variance. Paired
invariant tests are unaffected (the offset is common to both cells, cancels).
"""

from __future__ import annotations

import functools
import json

from lib_py.report import ARTIFACTS

_CLIP = 3.0


@functools.lru_cache(maxsize=1)
def _coeffs() -> dict:
    return json.loads((ARTIFACTS / "ratings" / "rating_effect_coefficients.json").read_text())


@functools.lru_cache(maxsize=1)
def _ref() -> dict:
    return json.loads((ARTIFACTS / "ratings" / "attribute_reference_stats.json").read_text())[
        "attribute_reference_stats"
    ]


def families() -> dict:
    return _coeffs()["families"]


def _unit_mean(players: list[dict], attr: str) -> float | None:
    vals = [p["attributes"][attr] for p in players
            if p and isinstance(p.get("attributes"), dict) and p["attributes"].get(attr) is not None]
    return sum(vals) / len(vals) if vals else None


_mod_cache: dict = {}


def family_modifier(fam_key: str, players: list[dict], scale: float = 1.0) -> float:
    """Sum beta·z over the family's attributes for this set of participating players."""
    key = (fam_key, tuple(p["id"] if p else None for p in players))
    hit = _mod_cache.get(key)
    if hit is not None:
        return hit * scale
    fam = families()[fam_key]
    ref = _ref()
    total = 0.0
    for a in fam["attributes"]:
        attr = a["attribute"]
        m = _unit_mean(players, attr)
        r = ref.get(attr)
        if m is None or not r or not r.get("sd"):
            continue
        z = (m - r["mean"]) / r["sd"]
        z = max(-_CLIP, min(_CLIP, z))
        total += a["beta_per_z"] * z
    _mod_cache[key] = total
    return total * scale


# ---- §12 league centering -----------------------------------------------------
# Per-family mean modifier over the on-field slot across all 32 depth charts.
# Subtracting it makes an average real matchup produce ≈0 net shift.

@functools.lru_cache(maxsize=1)
def _offsets() -> dict[str, float]:
    from engine.roster import roster, team_list  # lazy: roster imports nothing here
    slots = {
        "qb_accuracy": lambda o, d: [o["QB1"]],
        "qb_ball_security": lambda o, d: [o["QB1"]],
        "receiver_hands_routes": lambda o, d: [o["WR1"], o["WR2"], o["WR3"], o["TE1"]],
        "yac_ballcarrier": lambda o, d: [o["WR1"]],
        "coverage": lambda o, d: [d["CB1"], d["CB2"], d["S1"], d["S2"]],
        "open_field_tackling": lambda o, d: [d["CB1"], d["CB2"], d["S1"], d["S2"], d["ILB1"], d["ILB2"]],
        "protection": lambda o, d: [o["LT"], o["LG"], o["C"], o["RG"], o["RT"]],
        "pass_rush": lambda o, d: [d["EDGE1"], d["EDGE2"], d["DT1"], d["DT2"]],
        "run_defense_front7": lambda o, d: [d["EDGE1"], d["EDGE2"], d["DT1"], d["DT2"], d["ILB1"], d["ILB2"]],
        "runner": lambda o, d: [o["RB1"]],
    }
    teams = team_list()
    off = {t: roster(t).offense() for t in teams}
    deff = {t: roster(t).defense() for t in teams}
    out: dict[str, float] = {}
    for fam, pick in slots.items():
        vals = [family_modifier(fam, pick(off[t], deff[t])) for t in teams]
        out[fam] = sum(vals) / len(vals)
    kx = [family_modifier("kicking", [roster(t).kicker()]) for t in teams]
    out["kicking"] = sum(kx) / len(kx)
    return out


def _centered(fam_key: str, players: list[dict]) -> float:
    return family_modifier(fam_key, players) - _offsets().get(fam_key, 0.0)


# ---- which families feed which resolver call (spec §11 player-modifier lists) --

def completion_logit_shift(off_pass_catchers, defenders, qb) -> float:
    """M09 COMPLETE: qb accuracy (+) + receiver hands/routes (+) + coverage (-)."""
    return (
        _centered("qb_accuracy", [qb])
        + _centered("receiver_hands_routes", off_pass_catchers)
        + _centered("coverage", defenders)
    )


def interception_logit_shift(qb) -> float:
    """M09 INTERCEPTION: qb ball security (-)."""
    return _centered("qb_ball_security", [qb])


def sack_logit_shift(ol, rushers) -> float:
    """M04 SACK: protection (-) + pass rush (+)."""
    return _centered("protection", ol) + _centered("pass_rush", rushers)


def rush_yards_shift(ol, front7, runner) -> float:
    """M14 yards: runner (+) + OL (folded into runner family weights) + front seven (-)."""
    return _centered("runner", [runner]) + _centered("run_defense_front7", front7)


def yac_yards_shift(receiver, tacklers) -> float:
    """M10 yards: ball-carrier (+) + open-field tackling (-)."""
    return _centered("yac_ballcarrier", [receiver]) + _centered("open_field_tackling", tacklers)


def fg_logit_shift(kicker) -> float:
    return _centered("kicking", [kicker])

File: src/engine/ratings.ts
/**
 * Rating layer (spec §12–§14) — turns lineups into logit / yardage modifiers.
 *
 * TS port of `analysis/engine/ratings.py`.
 *
 *   modifier(family) = Σ_attr  beta_per_z(attr) · clip(z(attr), -3, 3)
 *
 * z(attr) = (unit_mean_rating − ref_mean) / ref_sd over the participating
 * players who have that attribute. Signs are folded into `beta_per_z`.
 *
 * §12 centering: `ref_mean` is over the whole reference pool, but the engine
 * only fields starters (above that mean), so each `*Shift` subtracts the
 * family's league-mean modifier over its on-field slot (`offsets()`) — an
 * average real matchup then nets ≈0 and only matchup *differences* move the
 * outcome.
 */

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

import type { Player } from "../schema/player.js";
import { type Lineup, roster, teamList } from "./roster.js";

/** A "unit" is the set of participating players; nulls/undefined are ignored. */
type Unit = readonly (Player | null | undefined)[];

const CLIP = 3.0;

interface FamilyAttr {
  attribute: string;
  beta_per_z: number;
}
interface Family {
  attributes: FamilyAttr[];
}
interface RefStat {
  mean: number;
  sd: number;
}

function readArtifact<T>(rel: string): T {
  const url = new URL(`../../artifacts/ratings/${rel}`, import.meta.url);
  return JSON.parse(readFileSync(fileURLToPath(url), "utf8")) as T;
}

const FAMILIES: Record<string, Family> = readArtifact<{ families: Record<string, Family> }>(
  "rating_effect_coefficients.json",
).families;

const REF: Record<string, RefStat> = readArtifact<{
  attribute_reference_stats: Record<string, RefStat>;
}>("attribute_reference_stats.json").attribute_reference_stats;

/** Mean of `attr` over players that have it, or null. */
function unitMean(players: Unit, attr: string): number | null {
  let sum = 0;
  let n = 0;
  for (const p of players) {
    const v = p?.attributes?.[attr as keyof Player["attributes"]];
    if (typeof v === "number") {
      sum += v;
      n += 1;
    }
  }
  return n ? sum / n : null;
}

const _modCache = new Map<string, number>();

/** Σ beta·z over the family's attributes for this participating unit. */
export function familyModifier(famKey: string, players: Unit, scale = 1): number {
  const key = `${famKey}|${players.map((p) => p?.id ?? "").join(",")}`;
  const hit = _modCache.get(key);
  if (hit !== undefined) return hit * scale;

  const fam = FAMILIES[famKey];
  let total = 0;
  if (fam) {
    for (const a of fam.attributes) {
      const m = unitMean(players, a.attribute);
      const r = REF[a.attribute];
      if (m === null || !r || !r.sd) continue;
      let z = (m - r.mean) / r.sd;
      z = Math.max(-CLIP, Math.min(CLIP, z));
      total += a.beta_per_z * z;
    }
  }
  _modCache.set(key, total);
  return total * scale;
}

// ---- §12 league centering -------------------------------------------------

type Pick = (o: Lineup, d: Lineup) => Unit;

const SLOTS: Record<string, Pick> = {
  qb_accuracy: (o) => [o.QB1],
  qb_ball_security: (o) => [o.QB1],
  receiver_hands_routes: (o) => [o.WR1, o.WR2, o.WR3, o.TE1],
  yac_ballcarrier: (o) => [o.WR1],
  coverage: (_o, d) => [d.CB1, d.CB2, d.S1, d.S2],
  open_field_tackling: (_o, d) => [d.CB1, d.CB2, d.S1, d.S2, d.ILB1, d.ILB2],
  protection: (o) => [o.LT, o.LG, o.C, o.RG, o.RT],
  pass_rush: (_o, d) => [d.EDGE1, d.EDGE2, d.DT1, d.DT2],
  run_defense_front7: (_o, d) => [d.EDGE1, d.EDGE2, d.DT1, d.DT2, d.ILB1, d.ILB2],
  runner: (o) => [o.RB1],
};

let _offsets: Record<string, number> | null = null;

/** Per-family mean modifier over its on-field slot across all 32 depth charts. */
export function offsets(): Record<string, number> {
  if (_offsets) return _offsets;
  const teams = teamList();
  const off = new Map(teams.map((t) => [t, roster(t).offense()]));
  const def = new Map(teams.map((t) => [t, roster(t).defense()]));
  const out: Record<string, number> = {};
  for (const [fam, pick] of Object.entries(SLOTS)) {
    let s = 0;
    for (const t of teams) s += familyModifier(fam, pick(off.get(t)!, def.get(t)!));
    out[fam] = s / teams.length;
  }
  let sk = 0;
  for (const t of teams) sk += familyModifier("kicking", [roster(t).kicker()]);
  out.kicking = sk / teams.length;
  _offsets = out;
  return out;
}

function centered(famKey: string, players: Unit): number {
  return familyModifier(famKey, players) - (offsets()[famKey] ?? 0);
}

// ---- resolver shift functions (spec §11 player-modifier lists) -----------

/** M09 COMPLETE: qb accuracy (+) + receiver hands/routes (+) + coverage (−). */
export function completionLogitShift(
  catchers: Unit,
  defenders: Unit,
  qb: Player | null | undefined,
): number {
  return (
    centered("qb_accuracy", [qb]) +
    centered("receiver_hands_routes", catchers) +
    centered("coverage", defenders)
  );
}

/** M09 INTERCEPTION: qb ball security (−). */
export function interceptionLogitShift(qb: Player | null | undefined): number {
  return centered("qb_ball_security", [qb]);
}

/** M04 SACK: protection (−) + pass rush (+). */
export function sackLogitShift(ol: Unit, rushers: Unit): number {
  return centered("protection", ol) + centered("pass_rush", rushers);
}

/** M14 yards: runner (+) + front seven (−). */
export function rushYardsShift(
  _ol: Unit,
  front7: Unit,
  runner: Player | null | undefined,
): number {
  return centered("runner", [runner]) + centered("run_defense_front7", front7);
}

/** M10 yards: ball-carrier (+) + open-field tackling (−). */
export function yacYardsShift(receiver: Player | null | undefined, tacklers: Unit): number {
  return centered("yac_ballcarrier", [receiver]) + centered("open_field_tackling", tacklers);
}

/** M20 MADE: kicking (+). */
export function fgLogitShift(kicker: Player | null | undefined): number {
  return centered("kicking", [kicker]);
}
